src/NHibernate/Linq/LinqExtensionMethodAttribute.cs

```diff
@@ -31,4 +47,58 @@ public LinqExtensionMethodAttribute(string name) : this()
 		/// </summary>
 		public string Name { get; }
 	}
+
+	/// <summary>
+	/// Can flag a method as not being callable by the runtime, when used in Linq queries.
+	/// If the method is supported by the linq-to-nhibernate provider, it will always be converted
+	/// to the corresponding SQL statement.
+	/// Otherwise the linq-to-nhibernate provider evaluates method calls when they do not depend on
+	/// the queried data.
+	/// </summary>
+	public class NotEvaluatableAttribute : LinqExtensionMethodAttributeBase
+	{
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		public NotEvaluatableAttribute()
+			: base(LinqExtensionEvaluation.NoEvaluation) { }
+	}
+
+	/// <summary>
+	/// Base class for Linq extension attributes.
+	/// </summary>
+	public abstract class LinqExtensionMethodAttributeBase : Attribute
+	{
+		/// <summary>
+		/// Should the method call be partially evaluated?
+		/// </summary>
+		public LinqExtensionEvaluation Evaluation { get; }
+
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		/// <param name="evaluation">Should the method call be partially evaluated?.</param>
+		protected LinqExtensionMethodAttributeBase(LinqExtensionEvaluation evaluation)
+		{
+			Evaluation = evaluation;
+		}
+	}
+
+	/// <summary>
+	/// Possible method call behaviors when the linq to NHibernate provider partially evaluates
+	/// expressions before parsing them.
+	/// </summary>
+	public enum LinqExtensionEvaluation
+	{
+		/// <summary>
+		/// The method call will no be evaluated even if it arguments do not depend on the queried data.
+		/// </summary>
+		NoEvaluation,
+		/// <summary>
+		/// If the method call do not depend on queried data, the method call will be evaluated and replaced
+		/// by the resulting constant expression in the expression tree. This was the default behavior
+		/// prior to NHibernate v5.
+		/// </summary>
```

Comment:
I think this comment shall be reworded. First, need to remove mentions of the previous behavior, as it does not matter. Second, if method throws - it will throw.